Make storage engine configurable

david-molnar-oculai opened the pull request:
In mongodb v7 the ephemeral storage engine is not supported anymore. I made it configurable, so it can be disabled.

david-molnar-oculai committed:
Add support for storage engine override

david-molnar-oculai committed:
Set storage engine only when not none

david-molnar-oculai committed:
Drop warning

ekarademir commented:
Thank you for the PR @david-molnar-oculai ! Only comment I can make is, since we are basically opening up the possibility of writing data on disk, what are the possible ramifications? I guess, since data folder is configurable, it only makes sense to make this configurable?

When I first started this project, it was heavily influenced by the [mockgo](https://www.npmjs.com/package/mockgo) library in the JS world. I think `ephemeralForTest` is coming from that. Also it has been listed in the help text for the `mongod`. But I couldn't find any reference of it in the docs, going [back to 3.2](https://www.mongodb.com/docs/v3.2/core/inmemory/). I guess it was just for their [internal testing](https://jira.mongodb.org/browse/SERVER-65151).

Then maybe the fallback option should be `inMemory` rather than `ephemeralForTest`?

ekarademir commented on pymongo_inmemory/context.py:
```diff
@@ -127,6 +127,7 @@ def __init__(
         )
         self.archive_folder = mkdir_ifnot_exist(self.download_folder, self.url_hash)
         self.extracted_folder = mkdir_ifnot_exist(self.extract_folder, self.url_hash)
+        self.storage_engine = "ephemeralForTest"
```
Maybe the fallback option should be `inMemory`?

david-molnar-oculai replied:
Yes, I like `inMemory`. I'm not sure though what you mean by "fallback option". When would you apply the fallback option?
In this PR, the default setting (`ephemeralForTest`) doesn't change. It is just made configurable using the `Context` object.
If someone wants to not configure any specific storage engine or setting it to `inMemory`, that's now also possible. Of course, if you do this you need to consider the cases you mentioned (writing data to disk, permissions etc.).

david-molnar-oculai replied:
I tested it a bit with `inMemory`, but it doesn't work as it looks like this option only available in the Enterprise version (https://www.mongodb.com/docs/manual/core/inmemory/)
> Starting in MongoDB Enterprise version 3.2.6, the in-memory storage engine is part of general availability (GA) in the 64-bit builds.

david-molnar-oculai committed:
Use inMemory instead of ephemeralStorage by default

ephemeralStorage was removed from newer versions and it was meant to be used for internal testing

david-molnar-oculai committed:
Revert back to ephemeralForTest as inMemory only in Enterprise version

ekarademir commented:
Thank you for all the good work @david-molnar-oculai I think this is good to merge as is. By fallback, I meant default value for the option we are adding with this PR. But thank you for testing the `inMemory` availability.

ekarademir commented:
This is now published. Thank you for the contribution @david-molnar-oculai
